tests/config_schema_examples/slurm-example.yml

```diff
@@ -1,8 +1,13 @@
-
 editor: vi
 executors:
-  slurm:
+  normal:
     type: slurm
     description: submit jobs with slurm using srun instead of sbatch
+    launcher: sbatch
     vars:
-      launcher: srun
+      FOO: BAR
+    options:
```

Comment:
Where are options parsed, and why are there vars here if we removed them?

Reply:
If vars was in the example, why did tests still pass before?

Reply:
it's bit odd since ``additionalProperties: False`` in slurm object https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_schema/buildtest/settings/default.schema.json#L48 i have no idea

Reply:
I didn't parse ``options`` yet, that was something i plan to do in later PR. The #SBATCH directives would be read from the options key and injected into the test, so this would require testing on my end at stampede2.

Reply:
I think it's an issue with the nested ref, this looks like something that might work! https://stackoverflow.com/questions/58170135/how-to-use-additionalproperties-from-a-ref-file/58171051#58171051

Reply:
I looked at the link, it's bit confusing according to the docs https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-6.5.6 

This looks like a silent bug that doesn't show up during testing, and could be something to address in a separate PR. If i understand, we dont want any extra properties passed to any object   whether it be a child object like executor or the parent objects.

Reply:
Basically, the properties serve as booleans to reference if a definition is present or not (and according to the author) this fixes the bug that nesting of schemas doesn't honor the attribute. It's worth a try.

Reply:
so should i remove the following ``additionalProperties`` from ``local`` and ``slurm``. 
https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_schema/buildtest/settings/default.schema.json#L48
https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_schema/buildtest/settings/default.schema.json#L29

I guess the metric for success would be to get the validate method fail when ``vars`` is specified in executor object. Is that right?

Reply:
Yes we want the additionalProperties: false to be honored - in the case of vars it should have triggered an invalid error, but it didn't. It will be important beyond vars, because it could be that a user has a spelling mistake (or something like that) that should be known.

Reply:
Let me do some testing and see what i can fix.

Reply:
so i did a few tests

Test on outer level
---------------------
Expected Result: Fail on ``x: 1``
```
x: 1
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
```

```
On instance:
    {'editor': 'vi',
     'executors': {'runlocal': {'description': 'submit jobs locally',
                                'type': 'local'},
                   'x': {'type': 'local'}},
     'x': 1}
```
Test inside executor
-----------------------

Expected Result: Fail on ``vars: 1`` not an object 

```
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
  x:
    type: local
  vars: 1
```

```
Failed validating 'type' in schema[0]:
    {'additionalProperties': False,
     'properties': {'description': {'type': 'string'},
                    'options': {'items': {'type': 'string'},
                                'type': 'array'},
                    'type': {'enum': ['local'], 'type': 'string'}},
     'required': ['type'],
     'type': 'object'}

On instance:
    1
```

Test inside executor object
------------------------------

Expected Result: Fail on ``x: 1``

```
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
    x: 1
```

What i dont understand is why validate is thinking it needs to match with ``slurm`` object when a new key is being added. Maybe it resorts to the last object since it is using ``anyOf``

```
jsonschema.exceptions.ValidationError: 'local' is not one of ['slurm']

Failed validating 'enum' in schema[1]['properties']['type']:
    {'enum': ['slurm'], 'type': 'string'}

On instance['type']:
    'local'
```

I get same result when i change above schema from ``type: local`` to ``type: slurm``.

Now trying the combined schema from regtest but swap the two executor names and put ``vars: 1`` and it gets same result

```
editor: vi
executors:
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  local:
    type: local
    description: submit jobs locally
    vars: 1
```

```
jsonschema.exceptions.ValidationError: 'local' is not one of ['slurm']

Failed validating 'enum' in schema[1]['properties']['type']:
    {'enum': ['slurm'], 'type': 'string'}

On instance['type']:
    'local'
```

Now this last one is a bit interesting, apparently I can define two executors of same name ``local`` and the validate passes

```
editor: vi
executors:
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  local:
    type: local
    description: submit jobs locally

  local:
    type: slurm
    launcher: sbatch
    description: submit jobs locally                                      
```

For now i removed the two lines as i referenced in previous message.

Reply:
all my test were conducted with the following commit https://github.com/HPC-buildtest/buildtest-framework/pull/249/commits/3c1b5d8a7386d0aaeedf059e9523f6112bcbc299

Reply:
@vsoch what your thoughts regarding my results? Do you think the issue is resolved?

Reply:
I don't see a case where you add additional variables to an executor and have the expected failure, so I don't think so. If you want me to give it a shot I can make time for it this week.

Reply:
I think that would be good use-case of testing bad configuration examples as part of regression test. Maybe put them in this directory https://github.com/HPC-buildtest/buildtest-framework/tree/devel/tests/config_schema_examples 

that case that can be scoped for a new PR.

Reply:
Yes, but the issue we were addressing is having undefined fields in the executors, you asked me if that is resolved, and it is not.

Reply:
@shahzebsiddiqui for the last case, when you load the schema it will actually only have *one* of those executors because it's read as a dictionary. There might be a way to prevent duplicate keys but I haven't found it yet - we couldn't check for it in python because it would only load one of the sections.

Reply:
If I may let's document and handle testing of configuration check in separate issue. It's out of scope and not the intent when I was issuing the PR.